feat(iota-core, iota-types): Add scoring function

oliviasaa opened the pull request:
```mermaid
---
config:
  gitGraph:
    {parallelCommits: true,
    mainBranchName: 'Feature',
    rotateCommitLabel: true}
---
gitGraph
   commit id: "Initial commit"
   commit id: "Old Scorer" tag: "PR8530"
  commit id: "add ChangeEpochV4" tag: "PR9159"
    commit id: "add MisbehaviorReports" tag: "PR9175"
   commit id: "add ScoringMetrics" tag: "PR9176"
   commit id: "add Scorer" tag: "PR9177"
   commit id: "add ScoringMetricsStore" tag: "PR9178"
commit id: "add report validation" tag: "PR9362"
   checkout Feature
branch scoring-function
commit id: "add scoring function" 
branch report-creation
commit id: "add report creation" 
checkout Feature
merge  scoring-function tag: "THIS PR"
```
# Description of change

We Introduce the scoring formulas used for the validator report mechanism, which where only placeholder before this PR.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [x] Protocol: Introduce new scoring formulas for the validator score mechanism
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

oliviasaa committed:
Add report validation logic

oliviasaa committed:
Add scoring function

oliviasaa committed:
Add score update logic

oliviasaa committed:
fix typo

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -11,17 +11,30 @@ use iota_types::{
     pub(crate) current_local_metrics_count: Arc<VersionedScoringMetrics>,
+    // The metrics counts received from other authorities, i.e., the information contained in the
+    // MisbehaviourReports received by the authority. If an authority has not sent a report, its
+    // received_metrics will be all zeroed.
```
```suggestion
    // entry in this vector will be all zeroed.
```

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -30,45 +43,67 @@ impl Scorer {
                         })
                         .collect();
+                let parameters = ParametersV1 {
+                    max_score,
```
Can we just use `max_score` here in place of `2_u64.pow(16)` in all these params? If we changed `max_score` would these all change too? Or is it a separate value?

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
-        // Placeholder
+        // Vector with the highest received reports from each authority and their voting
+        // power. Authorities that did not send reports are filtered out.
+        let highest_received_reports_from_authority = self
```
What does `highest` refer to? Should it be `latest` instead, indicating that we use the most recent received report from each authority? `highest` could imply we take the max value or something.

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+            .collect::<Vec<(VersionedMisbehaviorReport, VotingPower)>>();
+        // Ensure that we have at least one report to calculate the scores, otherwise we
+        // do nothing.
+        if highest_received_reports_from_authority.is_empty() {
+        } else {
```
```suggestion
        if !highest_received_reports_from_authority.is_empty() {
```

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+fn calculate_scores_v1(median_reports: Vec<MedianMetricVec>, parameters: ParametersV1) -> Vec<u64> {
+    // Ensure we have 4 median reports: faulty blocks provable, faulty blocks
+    // unprovable, missing blocks, equivocations. This is naturally guaranteed when
+    // these data come from MisbehaviorReports, since they would been considered
```
```suggestion
    // these data come from MisbehaviorReports, since they would have been considered
```

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+    // process.
+    assert!(!reports_and_voting_power.is_empty());
+
+    let number_of_metrics = reports_and_voting_power[0].0.iterate_over_metrics().len();
```
define a constant `number_of_metrics_v1` because you use the magic number 4 below to check this as well.

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+    // unprovable, missing blocks, equivocations. This is naturally guaranteed when
+    // these data come from MisbehaviorReports, since they would been considered
+    // invalid otherwise.
+    assert_eq!(median_reports.len(), 4);
```
A suggested above, define a constant for the number of metrics in v1 instead.

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -30,45 +43,67 @@ impl Scorer {
+                    weights: vec![
+                        2_u64.pow(16) * 30 / 100,
+                        2_u64.pow(16) * 10 / 100,
+                        2_u64.pow(16) * 35 / 100,
```
Will these remain hardcoded here or be defined elsewhere in another PR? If they remain here it would be worth commenting here which values correspond to which metrics and how they should be interpreted.

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -30,45 +43,67 @@ impl Scorer {
+                    allowances: vec![1, 2, 1000, 0],
+                    maximums: vec![5, 10, 5000, 1],
+                    weights: vec![
+                        2_u64.pow(16) * 30 / 100,
```
These are `scale_factor`, right? Should define it above and use it here.

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+        parameters.scale_factor,
+        parameters.max_score,
+    )
+}
```
I actually think it might be better not to group all the allowances, maximums and weights into a vector, it just end up being more confusing. It will take more lines of code, but I think it would be better to just explicitly list them all, `missing_allowance`, etc...

cyberphysic4l commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +112,301 @@ impl Scorer {
+                .zip(to_add.iter())
+                .for_each(|(a, b)| *a += *b);
+        });
+    // Combine equivocation scores into final scores, and scale down. The final
```
Mention that equivocation score is between zero and max_score too, which is why we are dividing by max_score.

cyberphysic4l reviewed (changes requested):
My main comment is just about the scoring function and the iterating over the metrics to calculate the separate scores. My feeling is that it might be better to simply enumerate all the metrics and the're parameters in the scoring function by name for the sake of clarity, even if this means more lines of code. I feel we don't get much benefit from having them in a vector. We can discuss this further.

cyberphysic4l committed:
Use generic MisbehaviorsV1 struct for parameters and scores for each misbehavior metric

cyberphysic4l committed:
clippy and typos

bingyanglin commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -143,24 +487,232 @@ mod tests {
+                10_u64,
+            ),
+            (
+                VersionedMisbehaviorReport::V1(MisbehaviorsV1 {
+                    faulty_blocks_provable: vec![7, 8, 9],
+                    faulty_blocks_unprovable: vec![10, 11, 12],
+                    missing_proposals: vec![4, 5, 6],
+                    equivocations: vec![1, 2, 0],
+                }),
```
This setting exist multiple times, maybe it's cleaner to define something like `default_test_parameters() -> ParametersV1` so as to reduce the lines of code, and easier to maintain?

bingyanglin commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -77,37 +143,315 @@ impl Scorer {
+    } else if value >= max {
+        0
+    } else {
+        // Add overflow checks
```
```suggestion
        // TODO: add overflow checks
```
Also it might be good to define a corresponding issue for the TODO if you haven't PRed and/or planned.

cyberphysic4l replied:
Done, issue created https://github.com/iotaledger/iota-private/issues/242

bingyanglin commented on crates/iota-core/src/authority/scorer.rs:
```diff
@@ -30,45 +44,97 @@ impl Scorer {
                 ));
 
-                let (received_metrics, metrics_missing_from, current_scores, invalid_reports_count) =
+                let max_score = 2_u64.pow(16);
```
Would it be better to define a const like 
`const MAX_SCORE: u64 = 2_u64.pow(16);` in the front of the code? Same for the scale_factor.

oliviasaa committed:
feat(iota-core, iota-types): Enable misbehavior report creation (#9363)

# Description of change

This PR enables the creation of the MisbehaviorReport consensus message,
as their submission to consensus in order to disseminate it with the
rest of the committee.

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes #(issue)`.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

Check each box that your changes affect. If none of the boxes relate to
your changes, release notes aren't required.

For each box you select, include information after the relevant heading
that describes the impact of your changes that a user might notice and
any actions they must take to implement updates.

- [x] Protocol: Enables misbehavior report creation and dissemination.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

---------

Co-authored-by: Andrew <i.am.andrew.cullen@gmail.com>

oliviasaa committed:
feat(iota-core, iota-types): Add advance epoch v4 (#9424)

# Description of change

Enables the creation of a new version of advance epoch transaction and
new advance epoch function in iota-framework to modify rewards based on
scores. The existing validator reporting system remains whereby rewards
can be slashed entirely when a validator is reported. Now we
additionally modulate the rewards with the score provided as an argument
to the new version of the advance epoch function.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [x] Protocol: Enables the creation of a new version of advance epoch
transaction and new advance epoch function in iota-framework to modify
rewards based on scores.
- [ ] Nodes (Validators and Full nodes):
- [x] Indexer: Enables a new `ChangeEpochV4` system transaction which
may appear in checkpoints for networks where `score_based_rewards`
protocol config is enabled. Consequently **a dependency bump is
necessary for `iota_data_ingestion_core` library consumers (custom
indexers) to prevent outages**.
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

---------

Co-authored-by: Andrew <i.am.andrew.cullen@gmail.com>

piotrm50 commented on crates/iota-core/src/checkpoints/checkpoint_output.rs:
```diff
@@ -123,6 +123,20 @@ impl<T: SubmitToConsensus + ReconfigurationInitiator> CheckpointOutput
+        // misbehavior is detected, but we choose to send the ones that include only
+        // unprovable counts at this point, due to periodicity reasons and to ensure a
+        // (approximate) synchronization with the score updates.
+        let misbehavior_report = epoch_store.scorer.current_local_metrics_count.to_report();
```
So basically all validators send the report after each checkpoint they create? Do I understand correctly?

cyberphysic4l committed:
User constants for max score and scale factor

cyberphysic4l committed:
Merge branch 'protocol-research/feat/score-integration' into protocol-research/feat/score-integration-scoring-function
